akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSpec.scala

```diff
@@ -0,0 +1,137 @@
+import akka.stream.testkit.{ StreamSpec, TestSubscriber }
+import akka.stream.testkit.Utils.assertAllStagesStopped
+
+class FutureFlattenSpec extends StreamSpec {
+  val materializer = ActorMaterializer()
+  //implicit def ec: ExecutionContext = materializer.executionContext
+
+  "Future source" must {
+    "use default materializer" when {
+      implicit def m = materializer
+      implicit def ec = materializer.executionContext
+
+      "flattening elements" in assertAllStagesStopped {
+        val subSource: Source[Int, String] =
+          Source(List(1, 2, 3)).mapMaterializedValue(_ ⇒ "foo")
+
+        val futureSource = new FutureFlattenSource(Future(subSource))
+        val source: Source[Int, Future[String]] = Source.fromGraph(futureSource)
+
+        val materialized = Promise[String]()
+        val watched: Source[Int, NotUsed] = source.watchTermination() { (m, d) ⇒
+          materialized.completeWith(d.flatMap(_ ⇒ m))
+          NotUsed
+        }
+
+        val p = watched.runWith(Sink.asPublisher(false))
+        val c = TestSubscriber.manualProbe[Int]()
+        p.subscribe(c)
+
+        val sub = c.expectSubscription()
+        sub.request(5)
+
+        c.expectNext(1)
+        c.expectNext(2)
+        c.expectNext(3)
+
+        c.expectComplete()
+
+        Await.result(materialized.future, 3.seconds) should ===("foo")
+      }
+    }
+
+    "use a materializer without auto-fusing" when {
+      implicit val noFusing = ActorMaterializer(
+        ActorMaterializerSettings(system).withAutoFusing(false))
+      implicit def ec = noFusing.executionContext
+
+      val tooDeepForStack = 50000
+
+      // Seen tests run in 9-10 seconds, these test cases are heavy on the GC
+      val veryPatient = Timeout(20.seconds)
+
+      "flattening from a future graph" in assertAllStagesStopped {
+        val g = Source.fromFutureGraph(Future {
+          Thread.sleep(2000)
+          Fusing.aggressive((1 to tooDeepForStack).
+            foldLeft(Source.single(42).mapMaterializedValue(_ ⇒ 1))(
+              (f, i) ⇒ f.map(identity)))
+        })
+
+        val (mat, fut) = g.toMat(Sink.seq)(Keep.both).run()
+        mat.futureValue(veryPatient) should ===(1)
+        fut.futureValue(veryPatient) should ===(List(42))
+      }
+
+      "flattening from a completion stage" in assertAllStagesStopped {
+        val future: Future[Graph[SourceShape[Int], Int]] = Future {
+          Fusing.aggressive((1 to tooDeepForStack).
+            foldLeft(Source.single(43).mapMaterializedValue(_ ⇒ 1))(
+              (f, i) ⇒ f.map(identity)))
+        }
+        val stage: CompletionStage[Graph[SourceShape[Int], Int]] = future.toJava
+        val g = Source.fromGraphCompletionStage(stage)
+
+        val (mat, fut) = g.toMat(Sink.seq)(Keep.both).run()
+        mat.toScala.futureValue(veryPatient) should ===(1)
+        fut.futureValue(veryPatient) should ===(List(43))
+      }
+    }
+
+    // TODO: downstream cancels before the future is completed
```

Comment:
This first one could be tested by creating the `FutureFlattenSource` with a `Future` that you never complete, and run it with a `Sink.cancel`

Future completed with failure should be pretty easy

"downstream applying backpressure" can be done by creating with a future that is not completed and a probe downstream, then not emit any demand and complete the future, and after that emit demand and see that you get the first element

the last one could be done using a `Future(Source.single(whatever).mapMaterializedValue(_ => throw exception)))`